opttoy2/opttoy.go

```diff
@@ -0,0 +1,1165 @@
+	lShiftOp:            scalar,
+	rShiftOp:            scalar,
+	unaryPlusOp:         scalar,
+	unaryMinusOp:        scalar,
+	unaryComplementOp:   scalar,
+}
+
+var comparisonOpMap = [...]operator{
+	parser.EQ:                eqOp,
+	parser.LT:                ltOp,
+	parser.GT:                gtOp,
+	parser.LE:                leOp,
+	parser.GE:                geOp,
+	parser.NE:                neOp,
+	parser.In:                inOp,
+	parser.NotIn:             notInOp,
+	parser.Like:              likeOp,
+	parser.NotLike:           notLikeOp,
+	parser.ILike:             iLikeOp,
+	parser.NotILike:          notILikeOp,
+	parser.SimilarTo:         similarToOp,
+	parser.NotSimilarTo:      notSimilarToOp,
+	parser.RegMatch:          regMatchOp,
+	parser.NotRegMatch:       notRegMatchOp,
+	parser.RegIMatch:         regIMatchOp,
+	parser.NotRegIMatch:      notRegIMatchOp,
+	parser.IsDistinctFrom:    isDistinctFromOp,
+	parser.IsNotDistinctFrom: isNotDistinctFromOp,
+	parser.Is:                isOp,
+	parser.IsNot:             isNotOp,
+	parser.Any:               anyOp,
+	parser.Some:              someOp,
+	parser.All:               allOp,
+}
+
+var binaryOpMap = [...]operator{
+	parser.Bitand:   bitandOp,
+	parser.Bitor:    bitorOp,
+	parser.Bitxor:   bitxorOp,
+	parser.Plus:     plusOp,
+	parser.Minus:    minusOp,
+	parser.Mult:     multOp,
+	parser.Div:      divOp,
+	parser.FloorDiv: floorDivOp,
+	parser.Mod:      modOp,
+	parser.Pow:      powOp,
+	parser.Concat:   concatOp,
+	parser.LShift:   lShiftOp,
+	parser.RShift:   rShiftOp,
+}
+
+var unaryOpMap = [...]operator{
+	parser.UnaryPlus:       unaryPlusOp,
+	parser.UnaryMinus:      unaryMinusOp,
+	parser.UnaryComplement: unaryComplementOp,
+}
+
+// expr is a unified interface for both relational and scalar expressions in a
+// query. Expressions have optional inputs, projections and
+// filters. Additionally, an expression maintains a bitmap of required input
+// variables and a bitmap of the output variables it generates.
+//
+// Every unique column and every projection (that is more than just a pass
+// through of a variable) is given a variable index with the query. The
+// variable indexes are global to the query (see queryState) making the bitmaps
+// easily comparable. For example, consider the query:
+//
+//   SELECT x FROM a WHERE y > 0
+//
+// There are 2 variables in the above query: x and y. During name resolution,
+// the above query becomes:
+//
+//   SELECT @0 FROM a WHERE @1 > 0
+//   -- @0 -> x
+//   -- @1 -> y
+//
+// This is akin to the way parser.IndexedVar works except that we're taking
+// care to make the indexes globally unique. Because each of the relational
+// expression nodes maintains a bitmap of the variables it outputs we can
+// quickly determine if a scalar expression can be handled using bitmap
+// intersection.
+//
+// For scalar expressions the input variables bitmap allows an easy
+// determination of whether the expression is constant (the bitmap is empty)
```

Comment:
New paragraph:
"""
The bitmap determines which scalars are used at each level of a logical plan but it does not determine the order in which the values are presented in memory, e.g. as a result row. For this each expression must also carry, next to the bitmap, an (optional) *reordering array* which maps the positions in the result row to the indexes in the bitmap. For example, the two queries `SELECT k,v FROM kv` and `SELECT v,k FROM kv` have the same bitmap, but the first reorders `@1 -> idx 0, @2 -> idx 1` whereas the second reorders `@1 -> idx 1, @2 -> idx 0`. A subsequent RFC is to determine whether a single array is sufficient for this (indexed by the output column position) or whether both backward and forward associations must be maintained.
"""

Reply:
I think with this we do not need a separate rename stage.

Reply:
Interesting. This is effectively a rename stage, though baked into every node. I think you might be correct that this is necessary. I've added your paragraph as a TODO, sketching out this idea.

Reply:
During the initial construction of the expression this rename will be indeed present at every stage, because that's what SQL allows.

But perhaps we do not need it *embedded* in every stage, and instead have the initial construction "weave" it along during the recursion and resolve it at every stage, keeping the rename only for the end.

Although this is possible, I predict that the construction of the final physical plans will need to compute a rename at every stage where the entries in the bitmap are not contiguous and starting with position 0; so as to compact the data structure that represents the values in memory.  But I would agree it's too early to say.